internal/controller/controllers/bmh_agent_controller.go

```diff
@@ -626,6 +626,35 @@ func (r *BMACReconciler) findInstallationDiskID(devices []aiv1beta1.HostDisk, hi
 	return ""
 }
 
+// Finds the agents related to this ClusterDeployment
+//
+// The ClusterDeployment <-> agent relation is one-to-many.
+// This function returns all Agents whose ClusterDeploymentName name
+// matches this ClusterDeployment's name.
+func (r *BMACReconciler) findAgentsByClusterDeployment(ctx context.Context, clusterDeployment *hivev1.ClusterDeployment) []*aiv1beta1.Agent {
+	agentList := aiv1beta1.AgentList{}
+	err := r.Client.List(ctx, &agentList, client.InNamespace(clusterDeployment.Namespace))
+	if err != nil {
+		return nil
+	}
+
+	agents := []*aiv1beta1.Agent{}
+	for i, agent := range agentList.Items {
+		if agent.Spec.ClusterDeploymentName == nil {
+			continue
+		}
+		if agent.Spec.ClusterDeploymentName.Name == clusterDeployment.Name {
+			agents = append(agents, &agentList.Items[i])
+		}
```

Comment:
I believe there's a chance this will catch also "obsolete" agents. Agents with failed installations for this cluster deployment. 

In the `findAgents` function only the "newest" agent is returned: https://github.com/openshift/assisted-service/blob/a23228cab12c113b2a358a1d68a5a8a961e50992/internal/controller/controllers/bmh_agent_controller.go#L694-L698

I guess it's a bit tricky here because there could be more than one active agents per `ClusterDeployment` depending on how many nodes are being deployed.  One option could be to just do the queuing for agents that have the `AGENT_BMH_LABEL` and just get the newest of each set. 

Or do a bunch of queries to the kubernetes API and filter based on BMH,

Reply:
Good point. I've changed this function to return a single Agent per BMH. Uses a Set/Map as you suggested to remove duplicates by CreationTimestamp.